core/psh/psh.c

```diff
@@ -92,22 +92,24 @@ int main(int argc, char **argv)
 		usleep(50000);
 
 	/* Check if its first shell */
-	pid = tcgetpgrp(STDIN_FILENO);
-
+	tcpid = tcgetpgrp(STDIN_FILENO);
+	base = basename(argv[0]);
 	do {
 		/* login prompt */
-		base = basename(argv[0]);
 		if (strcmp(base, "pshlogin") == 0 && (app = psh_findapp("auth")) != NULL)
-			while (app->run(0, NULL) != 1)
+			while (app->run(0, NULL) != 0)
 				;
 
 		/* Run app */
-		if ((app = psh_findapp(base)) != NULL)
+		if ((app = psh_findapp(base)) != NULL) {
 			app->run(argc, argv);
-		else
+		}
+		else {
 			fprintf(stderr, "psh: %s: unknown command\n", argv[0]);
+			break;
+		}
 
-	} while (pid == -1);
+	} while (tcpid == -1 && strcmp(base, "pshlogin") == 0);
```

Comment:
seems wasteful to do strcmp every time, can probably be optimized in `isrunpsh` PR.